neps/nep-0491.md

```diff
@@ -0,0 +1,359 @@
+refund problems is hard.
+
+One proposed design is a combination of zero balance accounts and code sharing
+between contracts. Basically, if somehow the deployed code is stored in a way
+that does not require storage staking by the user themself, maybe the per-user
+state is small enough to fit in the 770 bytes limit of zero balance accounts.
+(Questionable for non-trivial use cases.)
+
+This alternative is much harder to design and implement. The proposal that has
+gotten the furthest so far is [Ephemeral
+Storage](https://github.com/near/NEPs/pull/485), which is pretty complicated and
+does not have community consensus yet. Nobody is currently working on moving it
+forward. While we could wait for that to eventually make progress, in the
+meantime, the community is held back in their innovation because of the refund
+problem.
+
+## Future possibilities
+
+- We might want to add the possibility to make non-refundable balance transfers
+  from within a smart contract. This would require changes to the WASM smart
+  contract to host interface. Since removing anything from there is virtually
+  impossible, we shouldn't be too eager in adding it there but if there is
+  demand for it, we certainly can do it without much trouble.
+- We could later add the possibility to refund the non-refundable tokens to the
+  account who sent the tokens initially.
+- We could allow sending non-refundable balance to existing accounts.
+- If (cheap) code sharing between contracts is implemented in the future, this
+  proposal will most likely work well in combination with that. Per-user data
+  will still need to be paid for by the user, which could be sponsored as
+  non-refundable balance without running into refund abuse.
+
+
+## Consequences
+
+
+### Positive
+
+- Businesses can sponsor new user accounts without the user being able to steal
+  any tokens.
+
+### Neutral
+
+- Non-refundable tokens are removed from the circulating supply, i.e. burnt.
+
+### Negative
+
+- Understanding a user's balance become even more complicated than it already
+  is. (Instead of only `amount` amd `locked`, there will be a third component.)
+- There is no incentive anymore to delete an account and its state when the
+  backing tokens are not refundable.
+
+
+### Backwards Compatibility
+
+We believe this can be implemented with full backwards compatibility.
+
+## Unresolved Issues (Optional)
+
+All of these issues already have a proposed solution above. But nevertheless,
+these points are likely to be challenged / discussed:
+
+- Should we allow adding non-refundable balance to existing accounts? (proposal:
+  no)
+- Should we allow adding more non-refundable balance after account creation?
+  (proposal: no)
+- Should this NEP include a host function to send non-refundable balance from
+  smart contracts? (proposal: no)
+- How should a wallet display non-refundable balances? (proposal: up to wallet
+  providers, probably a new separate field)
```

Comment:
Do wallets even need to display non-refundable balance to users (what for) ? My impression was "non-refundable balance" is working specifically towards hiding the complexity of storage-staking from users, is this even an achievable goal ? Blockchain explorers would probably need to display "non-refundable balance" though at least for info/debugging purposes.

On a similar topic, the name "non-refundable balance" might be unnecessarily vague given the intent of this proposal, wouldn't something like "dedicated storage balance" or "storage-tied balance" better describe the concept ? Since it's gonna show up in blockchain explorers this makes it a bit more self-explanatory, I think. Unless it does become a more generic concept eventually (in which case more generic name is warranted ofc).

Reply:
Hi @norwnd , thanks for your input, it's very much welcome! :)

> Do wallets even need to display non-refundable balance to users (what for) ?

Not displaying it is a valid choice for a wallet IMO. But right now, I assume most wallets are doing some kind of computation to show how much of the balance is currently open for withdrawal, vs the amount that's reserved to cover storage staking. If they have this computation, they need to update how it works once this proposal gets through.

> My impression was "non-refundable balance" is working specifically towards hiding the complexity of storage-staking from users, is this even an achievable goal ?

Interesting question. The original motivation here isn't so much to simplify things. It's more about enabling a certain use case. But I can see how, if presented correctly, we can maybe make it simpler from a user point of view, as they don't even have to know there is storage staking for their account. I like that!

> wouldn't something like "dedicated storage balance" or "storage-tied balance"/"storage-restricted balance" better describe the concept ?

I'm open to name changes. I very much dislike "non-refundable balance". But I would like to have a clear distinction between "refundable balance that's currently locked due to storage usage" (which we already have today) and the new concept of "balance that can only ever be used for storage in this account and will be burnt when the account is deleted". In your suggestions, it seems like the line between the two is blurry.

I've considered "storage credits", but that sounds like you can trade them between accounts.

Another idea: Perhaps we should just "purchase" an amount of bytes, and add a field like `free_bytes`, as in, no storage staking required for these bytes. Sounds a bit crazy to me but maybe that would make the concept easier to grasp. WDYT?

Reply:
> In your suggestions, it seems like the line between the two is blurry.

Yep, it was just to illustrate that point, your `free_bytes` concept feels much better (alternative names to consider: `purchased/prepayed/bought`+`_`+`bytes/storage` perhaps).

Reply:
> The original motivation here isn't so much to simplify things. It's more about enabling a certain use case. But I can see how, if presented correctly, we can maybe make it simpler from a user point of view, as they don't even have to know there is storage staking for their account.

Long term though, UX-wise both for **users** and **devs** the _"Alternative: Move away from storage staking"_ is still probably desirable, if achievable in sustainable manner.